### PR TITLE
KNOX-2946 - Cookie Path Scoping doesn't work when using default topology URL

### PR DESCRIPTION
**Pull request by zeroflag**
## What changes were proposed in this pull request?

When cookie path scoping is enabled knox will update the Path attribute of the cookies with `/gateway/topology_name`.
This doesn't work when a service is accessed via the default topology name (there is no `gateway/topology_name` in the url).

## How was this patch tested?

Enable cookie scoping + set default topology to sandbox:

```xml
    <property>
        <name>default.app.topology.name</name>
        <value>sandbox</value>
    </property>
    <property>
        <name>gateway.scope.cookies.feature.enabled</name>
        <value>true</value>
    </property>
```

Set service url in sandbox.xml:

```xml
    <service>
        <role>HIVE</role>
        <url>http://localhost:1701</url>
    </service>
```

Set up a test http server:

```smalltalk
Teapot on
    GET: '/' -> [:req |  
        ZnResponse noContent
    	      addCookie: ((ZnCookie name: 'test' value:'val') path: '/original/path'); 
    	      yourself 
         ];
	start. 
```

Unblock cookies in service.xml (data/services/hive/0.13.0/service.xml):
```xml
       <param>
            <name>responseExcludeHeaders</name>
            <value>SET-COOKIE;asdfg</value>
      </param>
```

Cookie path is updated when using the full URL:

```
$ curl -u admin:admin-password -k -v https://localhost:8443/gateway/sandbox/hive

< HTTP/1.1 204 No Content
< Set-Cookie: test=val; Path=/gateway/sandbox/original/path
```

Cookie path is not updated when using the default URL:

```
$ curl -u admin:admin-password -k -v https://localhost:8443/hive

< HTTP/1.1 204 No Content
< Set-Cookie: test=val; path=/original/path
```

### CODE REVIEW

File: gateway-provider-rewrite/src/main/java/org/apache/knox/gateway/filter/rewrite/api/CookieScopeServletFilter.java [hunk 1]
```diff
@@ -43,7 +45,12 @@ public void init( FilterConfig filterConfig ) throws ServletException {
   @Override
   protected void doFilter( HttpServletRequest request, HttpServletResponse response, FilterChain chain )
       throws IOException, ServletException {
-    chain.doFilter(request, new CookieScopeResponseWrapper(response, gatewayPath, topologyName));
+    if ("true".equals(request.getAttribute(DEFAULT_TOPOLOGY_FORWARD_ATTRIBUTE_NAME))) {
```
Comment (pzampino) [inline gateway-provider-rewrite/src/main/java/org/apache/knox/gateway/filter/rewrite/api/CookieScopeServletFilter.java:48]: Would Boolean.parseBoolean(request.getAttribute(DEFAULT_TOPOLOGY_FORWARD_ATTRIBUTE_NAME)) be better here?

File: gateway-provider-rewrite/src/main/java/org/apache/knox/gateway/filter/rewrite/impl/CookieScopeResponseWrapper.java [hunk 2]
```diff
@@ -46,8 +46,8 @@ public CookieScopeResponseWrapper(HttpServletResponse response, String gatewayPa
     public void addHeader(String name, String value) {
         if (SET_COOKIE.equals(name)) {
             String updatedCookie;
-            if (value.contains(COOKIE_PATH)) {
-                updatedCookie = value.replace(COOKIE_PATH, scopePath);
+            if (hasCookiePathAttribute(value)) {
+                updatedCookie = value.replaceAll("(?i)" + COOKIE_PATH, scopePath);
```
Comment (pzampino) [inline gateway-provider-rewrite/src/main/java/org/apache/knox/gateway/filter/rewrite/impl/CookieScopeResponseWrapper.java:50]: What is being done here?

**Review by pzampino — COMMENTED** [review]
Couple of questions...

File: gateway-provider-rewrite/src/main/java/org/apache/knox/gateway/filter/rewrite/api/CookieScopeServletFilter.java (hunk 1 above)
Comment (zeroflag) [reply gateway-provider-rewrite/src/main/java/org/apache/knox/gateway/filter/rewrite/api/CookieScopeServletFilter.java:48]: That might be a bit better. I'll modify it.

**Review by pzampino — APPROVED** [review]
LGTM

**Pull request merged**